### PR TITLE
extension: add ROS 2 Jazzy (core24)

### PR DESCRIPTION
**Pull request by artivis**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

### CODE REVIEW

**Comment by artivis** [comment]
The failing spread test is flutter and thus unrelated to this pr.

**Comment by artivis** [comment]
@Guillaumebeuzeboc review please?

**Review by Guillaumebeuzeboc — APPROVED** [review]

**Review by mr-cal — COMMENTED** [review]
Looks good, I have some minor feedback on the tests.

File: tests/unit/parts/plugins/test_colcon.py [hunk 1]
```diff
@@ -371,11 +449,131 @@ def test_get_build_commands_with_all_properties(
             f"--stage-cache-dir {new_dir} --base core22",
         ]
 
+        plugin = setup_method_fixture(
```
[the remark this answers is not in the document]
Comment (artivis) [reply tests/unit/parts/plugins/test_colcon.py:464]: Done.

File: tests/unit/parts/plugins/test_colcon.py [hunk 2]
```diff
@@ -112,22 +112,29 @@ def test_property_all(self):
...
 
         assert plugin.get_build_packages() == {
             "python3-colcon-common-extensions",
             "python3-rosdep",
-            "python3-rosinstall",
-            "python3-wstool",
             "rospack-tools",
         }
 
+        plugin = setup_method_fixture("core24", new_dir)
```
[the remark this answers is not in the document]
Comment (artivis) [reply tests/unit/parts/plugins/test_colcon.py:124]: Done.

File: tests/unit/parts/plugins/test_colcon.py [hunk 3]
```diff
@@ -231,11 +238,82 @@ def test_get_build_commands(self, setup_method_fixture, new_dir, monkeypatch):
             f"--stage-cache-dir {new_dir} --base core22",
         ]
 
+        plugin = setup_method_fixture("core24", new_dir)
```
[the remark this answers is not in the document]
Comment (artivis) [reply tests/unit/parts/plugins/test_colcon.py:248]: Done.

File: tests/spread/extensions/snaps/ros2-jazzy-hello/task.yaml
[the remark this answers is not in the document]
Comment (artivis) [reply tests/spread/extensions/snaps/ros2-jazzy-hello/task.yaml:1]: I did not, my bad. Removed.

File: snapcraft/extensions/ros2_jazzy.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    def is_experimental(base: Optional[str]) -> bool:
+        return True
+
+    @overrides
+    def get_root_snippet(self) -> Dict[str, Any]:
+        return {
+            "package-repositories": [
+                {
+                    "type": "apt",
+                    # "url": "http://packages.ros.org/ros2/ubuntu",
```
[the remark this answers is not in the document]
Comment (artivis) [reply snapcraft/extensions/ros2_jazzy.py]: At the time I opened this pr, the upstream project was still in the pre-release testing phase. It now has been released and I updated the repo url. It is thus unrelated to the 'experimental' flag of this plugin.

File: snapcraft/parts/plugins/colcon_plugin.py [hunk 5]
```diff
@@ -105,11 +105,7 @@ class ColconPlugin(_ros.RosPlugin):
 
     @overrides
     def get_build_packages(self) -> Set[str]:
-        return super().get_build_packages() | {
-            "python3-colcon-common-extensions",
-            "python3-rosinstall",
```
[the remark this answers is not in the document]
Comment (artivis) [reply snapcraft/parts/plugins/colcon_plugin.py:110]: Those packages aren't strictly necessary, but you're right that some users may have come to rely on them implicitely and thus their removal could break some existing snaps. I've reverted that for core22.

**Comment by artivis** [comment]
Not sure why the codespell CI job would fail on legacy stuff here. Should this be rebased maybe?

**Review by mr-cal — APPROVED** [review]
Thanks!

**Comment by mr-cal** [comment]
> Not sure why the codespell CI job would fail on legacy stuff here. Should this be rebased maybe?

Yeah, codespell got updated and started flagging more words.  It should be fixed with an update.

**Review by lengau — APPROVED** [review]

**Pull request merged**